### ISSUE TITLE
problem with choice tag

### ISSUE THREAD

**Issue by jamesb0nd** [wsdl2phpgenerator/wsdl2phpgenerator#277]
When there is a choice tag in the wsdl file as request parameter, let's say there are two elements under this choice tag:

<xsd:complexType name="XXXX">
        <xsd:sequence>
            <xsd:choice>
                <xsd:element name="A" type="A" minOccurs="0"></xsd:element>
                <xsd:element name="B" type="B" minOccurs="0"></xsd:element>
           </xsd:choice>
        </xsd:sequence>        
</xsd:complexType>

Then the generated class will have two properties with default value null. 
Class XXX{
      protected $A=null;
      protected $B=null
}

Now, if we choose to use the 2nd property when creating the webservice request object:
xxxx.setB('value');

The request object will be:
[$A=> null, $B=>'value'].

Then, when we call the soap client to call webservice:
$this->__soapCall('xxx', array($body));   // $body is the request object, which is [$A=> null, $B=>'value'].

The real soap request will be blank without correct output of $B, b/c in the wsdl file, $A and $B is under choice tag, only one of them can be part of the request. And as we give both $A and $B to soap client, although $A is null, soap client still uses $A and neglect $B.

<SOAP-ENV:Body><ns1:XXX></ns1:XXX></SOAP-ENV:Body>

PHP version is 5.6.1

**Comment by bennsel** [comment]
+1 I've the same Problem

**Comment by B4rb4ross4** [comment]
I have the following type:

`<xs:complexType name="VerkaufsparameterWert">
        <xs:choice>
            <xs:element name="datum" type="xs:date"/>
            <xs:element name="zeit" type="xs:time"/>
            <xs:element name="datumZeit" type="xs:dateTime"/>
            <xs:element name="text" type="v:String200"/>
            <xs:element name="ganzzahl" type="xs:long"/>
            <xs:element name="dezimalzahl" type="xs:decimal"/>
            <xs:element name="boolean" type="xs:boolean"/>
            <xs:element name="tkid" type="v:TKID"/>
            <xs:element name="adresse" type="v:Adresse"/>
            <xs:element name="nameVorname" type="v:NameVorname"/>
        </xs:choice>
    </xs:complexType>`

In the current version the generated class will have all attributes in the constructor as required fields:
`/**
     * @param date $datum
     * @param time $zeit
     * @param \DateTime $datumZeit
     * @param String200 $text
     * @param int $ganzzahl
     * @param float $dezimalzahl
     * @param boolean $boolean
     * @param TKID $tkid
     * @param Adresse $adresse
     * @param NameVorname $nameVorname
     */
    public function __construct($datum, $zeit, \DateTime $datumZeit, $text, $ganzzahl, $dezimalzahl, $boolean, $tkid, $adresse, $nameVorname)
    {
      $this->datum = $datum;
      $this->zeit = $zeit;
      $this->datumZeit = $datumZeit->format(\DateTime::ATOM);
      $this->text = $text;
      $this->ganzzahl = $ganzzahl;
      $this->dezimalzahl = $dezimalzahl;
      $this->boolean = $boolean;
      $this->tkid = $tkid;
      $this->adresse = $adresse;
      $this->nameVorname = $nameVorname;
    }`

That is something between not pretty and pretty bad. I had not yet the chance to check the problem mentioned above, but i guess it still exists. If it is solvable in a justifiable timeframe I'll fix it here at work and add a pull request.

**Comment by shr3k** [comment]
I think it's problem in SoapClient. I came across it too even in pure PHP implementation.  All that has helped is use local WSDL/XSD and remove the <xs:choice> in requests.

**Comment by boryashkin** [comment]
Yes, i've also had this problem, @shr3k. But i found a solution for that case (modifying xsd): change `<xs:choise>` to `<xs:choice minOccurs="0" maxOccurs="unbounded">`

**Comment by shr3k** [comment]
@boryashkin Thank you, it was very helpful. I will try to force our state administration to modify it too.

**Comment by ameenross** [comment]
Interesting problem. I ran into this today.

It's bothered me before that type classes have their properties predefined. It means it's not possible to distinguish between *omitted* and *empty*. Both will be an existing property with NULL value. Not normally a big issue, but combined with a stricter XSD (choice in this case) it can be.

I've solved it like this:

```php
/**
 * @property string $typeOfRemovableTop Omit the element entirely when the car has no removable top.
 */
class Car
{
    /**
     * @var string $brand
     */
    protected $brand;

    /**
     * @var string $model
     */
    protected $model;
}
```

Obviously the disadvantage of this solution is not being able to mark the (dynamic) property as private or protected, but I can live with that.
